common/achievement/evaluate.ts

```diff
@@ -47,21 +48,28 @@ export async function evaluateAchievement(
 
         const bucketCreatorFunction = bucketCreatorDefs[bucketCreator].function;
         const bucketAggregatorFunction = aggregators[bucketAggregator].function;
-        const aggFunction = aggregators[aggregator].function;
 
-        if (!bucketCreatorFunction || !bucketAggregatorFunction || !aggFunction) {
+        const aggregatorFunction = aggregators[aggregator].function;
+
+        if (!bucketCreatorFunction || !bucketAggregatorFunction || !aggregatorFunction) {
             logger.error(
                 `No bucket creator or aggregator function found for ${bucketCreator}, ${aggregator} or ${bucketAggregator} during the evaluation of achievement`
             );
             return;
         }
+        const bucketContext = await getBucketContext(relation);
 
-        const buckets = await bucketCreatorFunction(relation, recordValue);
-        const bucketEvents = createBucketEvents(eventsForMetric, buckets);
+        let buckets: BucketConfig;
+        if (!relation) {
+            buckets = { bucketKind: 'time', buckets: [] };
+        }
 
+        buckets = bucketCreatorFunction({ numberOfPeriod: recordValue, context: bucketContext });
```

Comment:
I would keep the name `recordValue` instead of `numberOfPeriod`

Reply:
I'd rather call it periodLength. Record value makes no sense since this variable points to the weeks / moths amout to create buckets for.

Reply:
There's a difference between a value's actual meaning and how we choose to use it. I'm not a fan of hiding this transition in a common place for all bucket creators. We are using the record value as periodLength, because it makes sense for our current use-cases, but what if there is another bucket creator that needs a different calculation for a period length?

If we decide to keep this approach, we should also shift the '+ 1' from the 'for' loops to this location and provide more documentation explaining why the 'record' value is always equal to 'numberOfPeriod'.